### PR TITLE
Allows AIs to set camera location hotkeys

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
AIs can use CTRL and a number key to set a location to that number key, and then press it later to jump their eye back to the same location. Tilde is also set to the AI's own core.

## Why It's Good For The Game
Lets the AI manage multiple areas of interest in an alternative way to using the multi-screen system.

## Changelog
:cl: Zxaber
add: AIs can now use CTRL and a number key (1 - 9) to assign a location to the number key, and then press the key later to jump their camera to that location. The numpad keys will work as aliases for their respective numbers.
add: The Tilde key, or zero (or numpad zero), will return you to the last point you jumped from.
/:cl:

Release your inner starcraft player.

### CODE REVIEW

**Comment by MrPerson** [comment]
Only people with 700+ APM can truly be called AI players tbh.

**Comment by zxaber** [comment]
I didn't think to test it, but there's a very good chance that numlock keys with numlock might show up as normal number keys, in which case it would already work.

**Comment by IllotAFV** [comment]
You already can change a limb, that you are targeting ,with numkeys. So, allowing key binds to camera location through numkeys sounds like a great way to go.

**Comment by zxaber** [comment]
So as it turns out, Numpad keys are not, in fact, read the same as the standard number keys. As such, adding the Numpad keys will end up being a lot more code.

Currently, in order to avoid a giant list of if statements, I'm abusing the fact that the number keys are, well, numbers. I index into a list using the number key itself as the index, grab the stored turf (if any), and that's that. But I can't exactly index into a list with "Numpad2".

So the question becomes will enough people appreciate the numpad keys being an option to warrant the extra code? Using them would require moving either a hand from the mouse, or else a hand from WASD (and these hotkeys do not work if the user isn't in hotkey mode).

**Comment by MrPerson** [comment]
Probably not worth it, although the code is trivial.

 
`

	var/static/list/numpad_mappings
	if(!numpad_mappings)
		numpad_mappings = list("numpad0" = "0", "numpad1" = "1", "numpad2" = "2", "numpad3" = "3", "numpad4" = "4", "numpad5" = "5", "numpad6" = "6", "numpad7" = "7", "numpad8" = "8", "numpad9" = "9", "divide" = "/", "multiply" = "*", "subtract" = "-", "add" = "+", "decimal" = ".")
	if(numpad_mappings[_key])
		_key = numpad_mappings[_key]
`

**Comment by zxaber** [comment]
Did it in two extra lines.

New Changes:
- Numpad buttons, when detected, are stripped of the "numpad" text, leaving behind only the number. Hence, "numpad1" is an alias for the top row "1" key, and will lead to the same camera hotkey. I don't know if anyone will use the numpad for this, but two lines is hardly worth worrying about.

- Jumping to a hotkey now saves the location you jumped from to a variable, and Tilde now returns you, in a sort of pseudo goto/return system. (Chaining jumps will just overwrite your return, I'm not going to emulate a stack) . I figure this is probably more useful than copying the functionality of the view_core hud button.

-Added a text to number conversion to fix numbers adding to the list the first time they're used rather than overwriting the placeholder null.

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 1]
```diff
@@ -162,6 +165,25 @@
...
+			_key = text2num(_key)
+			if(client.keys_held["Ctrl"]) //do we assign a new hotkey?
+				cam_hotkeys[_key] = eyeobj.loc
+				to_chat(src, "Location saved to Camera Group [_key].")
+				return
+			if(cam_hotkeys[_key]) //if this is false, no hotkey for this slot exists.
+				cam_prev = eyeobj.loc
+				eyeobj.setLoc(cam_hotkeys[_key])
+				return
+	. = ..()
```
Comment (ShizCalev) [inline code/modules/mob/living/silicon/ai/ai.dm]: return ..()

**Review by ShizCalev — DISMISSED** [review]

**Comment by kriskog** [comment]
>
>add: The Tilde key will return you to the last point you jumped from.

Just letting you know that the tilde character (and key position) is a terrible hotkey to use for anything ever, as it moves based on keyboard localization. While it's next to the "1" key on english keyboards, it requires a ctrl + alt modifier on a key next to my enter key.

Given that we cannot actually individually rebind keys here, I would recommend simply making "1" mapped to core.

**Comment by zxaber** [comment]
Man, that's _really_ unfortunate, because the tilde key spot is just _perfect_ for such a key.

I guess I'll look into alternatives. Out of curiosity, what do you have in that spot?

**Comment by kriskog** [comment]
I've got | in that position. Sadly there doesnt seem to be a standard for it, from what I can tell it's different for almost every localization.

**Comment by zxaber** [comment]
Yeah, I looked up various layouts, and it's a mess, for sure.

I'm been speaking with ActionNinja about a custom hotkey system, and it very well might be that one of us take it on very soon. With that in mind, I'd like to merge this PR mostly as-is, with the hope that a custom hotkey PR will be soon behind it.

In the meantime, I don't want to remove 1 from the available hotkeys, because it lessens the number of hotkeys easily within reach of WASD. I could set 0 as an alternative to tilde, though, since I'm not using it as a hotkey. Would that be usable for the time being?

**Comment by SouDescolado** [comment]
Lets say you are in your core (not saved), go to position 1(And do not move), go to position 2 then press the "go back" button. Would you go to the core or would you go back to position 1?

**Comment by ExcessiveUseOfCobblestone** [comment]
maybe i'm missing it but how do you fix a misclick or something similar?

**Comment by zxaber** [comment]
I'm lost, what do you mean by misclick?

**Comment by ExcessiveUseOfCobblestone** [comment]
I mean obv going to be hard to accidentally ctrl1 but what if you decide you want to use the hotkey for something else or, more likely, you mean to bind that camera to a different number?

Basically, how to rebind the key?

**Comment by zxaber** [comment]
You can just overwrite it. The Ctrl section is checked for first, and ends in a return, so you won't jump if Ctrl is held; it'll just rebind.

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 2]
```diff
@@ -162,6 +165,25 @@
...
+		if("`", "0")
+			if(cam_prev)
+				eyeobj.setLoc(cam_prev)
+			return
+		if("1", "2", "3", "4", "5", "6", "7", "8", "9")
+			_key = text2num(_key)
+			if(client.keys_held["Ctrl"]) //do we assign a new hotkey?
+				cam_hotkeys[_key] = eyeobj.loc
+				to_chat(src, "Location saved to Camera Group [_key].")
+				return
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/mob/living/silicon/ai/ai.dm]: tfw can't read

**Pull request merged**